runtime/perfetto/db.go

```diff
@@ -155,47 +143,269 @@ func (d *DB) Close() error {
+			return fmt.Errorf("write span attributes to %s: %w", d.fname, err)
+		}
+	}
+	return nil
 }
 
-func (d *DB) storeEncoded(ctx context.Context, app, version string, encoded []byte) error {
-	const stmt = `
-		INSERT INTO traces(app, version, events)
-		VALUES (?,?,?);
-	`
-	_, err := d.execDB(ctx, stmt, app, version, string(encoded))
+// TraceSummary stores summary information about a trace.
+type TraceSummary struct {
+	TraceID            string    // Unique trace identifier, in hex format.
+	StartTime, EndTime time.Time // Start and end times for the trace.
+
+	// TODO(spetrovic): Add more fields as we figure out what type of
+	// summary trace information we want to display.
+}
+
+// QueryTraces returns the summaries of the traces that match the given
+// query arguments, namely:
+//   - That have been generated by the given application version.
+//   - That fit entirely in the given [startTime, endTime] time interval.
+//   - Whose duration is in the given [durationLower, durationUpper] range.
+//   - Who are in the most recent limit of trace spans.
+//
+// Any query argument that has a zero value (e.g., empty app or version,
+// zero endTime) is ignored, i.e., it matches all spans.
+//
+// TODO(spetrovic): Add attribute matching.
+func (d *DB) QueryTraces(ctx context.Context, app, version string, startTime, endTime time.Time, durationLower, durationUpper time.Duration, limit int64) ([]TraceSummary, error) {
+	const query = `
+SELECT trace_id, start_time_unix_us, end_time_unix_us
+FROM spans
+WHERE
+	(parent_span_id="0000000000000000") AND
+	(app=? OR ?="") AND (version=? OR ?="") AND
+	(start_time_unix_us>=? OR ?=0) AND (end_time_unix_us<=? OR ?=0) AND
+	((end_time_unix_us - start_time_unix_us)>=? OR ?=0) AND
+	((end_time_unix_us - start_time_unix_us)<=? OR ?=0)
+GROUP BY trace_id
+ORDER BY end_time_unix_us DESC
+LIMIT ?
+`
+	var startTimeUs int64
+	if !startTime.IsZero() {
+		startTimeUs = startTime.UnixMicro()
+	}
+	var endTimeUs int64
+	if !endTime.IsZero() {
+		endTimeUs = endTime.UnixMicro()
+	}
+	durationLowerUs := durationLower.Microseconds()
+	durationUpperUs := durationUpper.Microseconds()
+	if limit <= 0 {
+		limit = math.MaxInt64
+	}
+	rows, err := d.queryDB(ctx, query, app, app, version, version, startTimeUs, startTimeUs, endTimeUs, endTimeUs, durationLowerUs, durationLowerUs, durationUpperUs, durationUpperUs, limit)
 	if err != nil {
-		return fmt.Errorf("write trace to %s: %w", d.fname, err)
+		return nil, err
 	}
-	return nil
+	defer rows.Close()
+	var traces []TraceSummary
+	for rows.Next() {
+		if err := rows.Err(); err != nil {
+			return nil, err
+		}
+		var traceID string
+		var startTimeUs, endTimeUs int64
+		if err := rows.Scan(&traceID, &startTimeUs, &endTimeUs); err != nil {
+			return nil, err
+		}
+		traces = append(traces, TraceSummary{
+			TraceID:   traceID,
+			StartTime: time.UnixMicro(startTimeUs),
+			EndTime:   time.UnixMicro(endTimeUs),
+		})
+	}
+	return traces, nil
```

Comment:
Shouldn't we check for rows.Err() at the end?

Reply:
Good point, my error checking code was wrong. Done.